datastore/storagebackend/postgresql/gettsattrgroups.go

```diff
@@ -0,0 +1,326 @@
+
+	for _, goName := range goNames {
+		field1 := tp1.Elem().FieldByName(goName)
+		if !field1.IsValid() {
+			return false, fmt.Errorf("tp1.Elem().FieldByName(%s) returned invalid value", goName)
+		}
+		field2 := tp2.Elem().FieldByName(goName)
+		if !field2.IsValid() {
+			return false, fmt.Errorf("tp2.Elem().FieldByName(%s) returned invalid value", goName)
+		}
+
+		if !field1.Equal(field2) {
+			return false, nil // at least one difference found
+		}
+	}
+
+	return true, nil // no differences found
+}
+
+// getCombo gets a version of tsMdata1 where only the given attributes (in goNames) have been set
+// (all other attributes having the default values).
+// Returns (TSMetadata, nil) upon success, otherwise (..., error).
+func getCombo(tsMdata1 *datastore.TSMetadata, goNames []string) (*datastore.TSMetadata, error) {
+	tsMdata2 := datastore.TSMetadata{}
+	tp1 := reflect.ValueOf(tsMdata1)
+	tp2 := reflect.ValueOf(&tsMdata2)
+
+	for _, goName := range goNames {
+		field1 := tp1.Elem().FieldByName(goName)
+		if !field1.IsValid() {
+			return nil, fmt.Errorf("invalid field (1) (goName: %s)", goName)
+		}
+		field2 := tp2.Elem().FieldByName(goName)
+		if !field2.IsValid() {
+			return nil, fmt.Errorf("invalid field (2) (goName: %s)", goName)
+		}
+		if !field2.CanSet() {
+			return nil, fmt.Errorf("unassignable field (2) (goName: %s)", goName)
+		}
+
+		switch field1.Kind() {
+		case reflect.String:
+			field2.SetString(field1.String())
+		default:
+			return nil, fmt.Errorf("unsupported type: %v (goName: %s)", field1.Kind(), goName)
+		}
+	}
+
+	return &tsMdata2, nil
+}
+
+// getTSAttrGroupsIncInstances populates groups from cols such that each group contains all
+// instances that match a unique combination of database values corresponding to cols.
+// All attributes, including those in cols, are set to the actual values found in the database.
+// Returns nil upon success, otherwise error.
+func getTSAttrGroupsIncInstances(
+	db *sql.DB, cols []string, groups *[]*datastore.TSMdataGroup) error {
+	allCols := getTSAllPBNames() // get all protobuf names of TSMetadata message
+
+	goNames, err := getTSGoNamesFromPBNames(cols)
+	if err != nil {
+		return fmt.Errorf("getTSGoNamesFromPBNames() failed: %v", err)
+	}
+
+	// query database for all columns in time_series, ordered by cols
+	allColsS := strings.Join(allCols, ",")
+	colsS := strings.Join(cols, ",")
+	query := fmt.Sprintf("SELECT %s FROM time_series ORDER BY %s", allColsS, colsS)
+	rows, err := db.Query(query)
+	if err != nil {
+		return fmt.Errorf("db.Query() failed: %v", err)
+	}
+	defer rows.Close()
+
+	// aggregate rows into groups
+	currInstances := []*datastore.TSMetadata{} // initial current instance set
+	for rows.Next() {
+		// extract tsMdata from current result row
+		tsMdata, err := scanTsMdata(rows, allCols)
+		if err != nil {
+			return fmt.Errorf("scanTsMdata() failed: %v", err)
+		}
+
+		if len(currInstances) > 0 { // check if we should create a new current instance set
```

Comment:
Put the comment on the next line, as now it seem like this line 237 does the check, while it is the next two lines.

Reply:
Here I disagree, but this is a matter of personal taste I guess. Note that the comment is _inside_ the block (i.e. to the right of the '{', so it applies to what goes on inside the block, not to the condition that caused the block to be executed. In the latter case, the comment should have been put on the line above the if-expression IMO. The reason the comment is put on the same line is to save some vertical space. Like when you say
```
x = .... // do ... to x
```
instead of
```
// do ... to x
x = ...
```
On the other hand, if this violates some official Go coding style, I can consider changing it. It's not such a big thing.

Reply:
I also tend to prefer this style:
```
if c { // do this
  ...
} else { // do that
  ...
}
```
over this one:
```
if c {
  // do this
  ...
} else {
  // do that
  ...
}
```